audio/pwaudio.c

```diff
@@ -14,26 +14,357 @@
+{
+/*    AudiodevPwOptions *popts = &dev->u.pw;
+*/
+    assert(dev->driver == AUDIODEV_DRIVER_PW);
+    return dev;
+}
+
+typedef struct PWVoiceOut {
+    HWVoiceOut  hw;
+    struct pw_stream *stream;
+    pwaudio *g;
+}
+PWVoiceOut;
+
+typedef struct PWVoiceIn {
+    HWVoiceIn  hw;
+    struct pw_stream *stream;
+    pwaudio *g;
+}
+PWVoiceIn;
+
+/*output data processing function to generate stuffs in the buffer*/
+static void on_process(HWVoiceOut *hw)
+{
+        PWVoiceOut *p = (PWVoiceOut *) hw;
+        struct pw_buffer *b;
+        struct spa_buffer *buf;
+        int n_frames, stride;
+
+        /*obtain a buffer to write into*/
+        if ((b = pw_stream_dequeue_buffer(p->stream)) == NULL) {
+                pw_log_warn("out of buffers: %m");
+                return;
+        }
+
+        /*Write data into buffer*/
+        buf = b->buffer;
+        if ((p = buf->datas[0].data) == NULL)
+                return;
+        stride = sizeof(float) * 2;
+        n_frames = SPA_MIN(b->requested, buf->datas[0].maxsize / stride);
+ 
+        buf->datas[0].chunk->offset = 0;
+        buf->datas[0].chunk->stride = stride;
+        buf->datas[0].chunk->size = n_frames * stride;
+
+        /*queue the buffer for playback*/
+        pw_stream_queue_buffer(p->stream, b);
+}
+
+static void on_stream_state_changed(HWVoiceOut *hw, void *_data, enum pw_stream_state old, enum pw_stream_state state,
+                                    const char *error)
+{
+    PWVoiceOut *p = (PWVoiceOut *) hw;
+    pwaudio *g = _data;
+
+    printf("stream state: \"%s\"\n", pw_stream_state_as_string(state));
+
+    switch (state) {
+    case PW_STREAM_STATE_ERROR:
+    case PW_STREAM_STATE_UNCONNECTED:
+        pw_main_loop_quit(g->thread_loop);
+        break;
+
+    case PW_STREAM_STATE_PAUSED:
+        printf("node id: %d\n", pw_stream_get_node_id(p->stream));
+        pw_loop_update_timer(pw_main_loop_get_loop(g->thread_loop),
+                        g->timer, NULL, NULL, false);
+        break;
+    case PW_STREAM_STATE_STREAMING:
+    {
+        struct timespec timeout, interval;
+
+        timeout.tv_sec = 0;
+        timeout.tv_nsec = 1;
+        interval.tv_sec = 0;
+        interval.tv_nsec = 40 * SPA_NSEC_PER_MSEC;
+
+        pw_loop_update_timer(pw_main_loop_get_loop(g->thread_loop),
+                        g->timer, &timeout, &interval, false);
+        break;
+    }
+    default:
+        break;
+    }
+}
+
+static void on_timeout(HWVoiceOut *hw, uint64_t expirations)
+{
+        PWVoiceOut *p = (PWVoiceOut *) hw;
+        pw_log_trace("timeout");
+        pw_stream_trigger_process(p->stream);
+}
+
+static const struct pw_stream_events stream_events = {
+        PW_VERSION_STREAM_EVENTS,
+        .process = on_process,
+        .state_changed = on_stream_state_changed,
+};
+
+static size_t qpw_write(HWVoiceOut *hw, void *data, size_t len)
+{
+    PWVoiceOut *pw = (PWVoiceOut *) hw;
+    pwaudio *c = pw->g;
+    const char *error = NULL;
+    size_t l;
```

Comment:
Do you need `l`? It is either 0 or unitialized, or not even used. Probably can be skipped.